app/src/server/utils/nodes/nodeProperties/monitorProperties.ts

```diff
@@ -0,0 +1,131 @@
+  beforeProcessing: async (node) => {
+    const { initialFilters, lastLoggedCallUpdatedAt, maxOutputSize, sampleRate } = node.config;
+
+    const numExistingEntries = await kysely
+      .selectFrom("NodeEntry as ne")
+      .where("ne.nodeId", "=", node.id)
+      .groupBy("ne.nodeId")
+      .select(sql<number>`count(*)::int`.as("count"))
+      .executeTakeFirst();
+
+    if (!numExistingEntries) return;
+
+    const sampleRateHash = calculateSampleRateHash(sampleRate);
+    const nodeId = node.id;
+
+    const loggedCallsToAdd = await constructLoggedCallFiltersQuery({
+      filters: initialFilters,
+      projectId: node.projectId,
+      baseQuery: kysely
+        .selectFrom("LoggedCall as lc")
+        .where("lc.updatedAt", ">=", dayjs(lastLoggedCallUpdatedAt).toDate()),
+    })
+      .leftJoin("NodeEntry as existingNe", (eb) =>
+        eb.onRef("existingNe.loggedCallId", "=", "lc.id").on("existingNe.nodeId", "=", node.id),
+      )
+      .where("existingNe.id", "is", null)
+      // hash lc.id and nodeId to compare against a sampleRate hash
+      .where(sql`md5(lc.id || '::' || ${nodeId}) > ${sampleRateHash}`)
+      .orderBy("lc.createdAt", "asc")
+      .limit(maxOutputSize - numExistingEntries.count)
+      .selectAll("lc")
+      .execute();
+
+    const entriesToImport = loggedCallsToAdd
+      .map((loggedCall) => {
+        try {
+          const tLoggedCall = typedLoggedCall(loggedCall);
+
+          const validated = validateRowToImport({
+            input: tLoggedCall.reqPayload,
+            output: tLoggedCall.respPayload?.choices?.[0]?.message,
+          });
+
+          if ("error" in validated) return null;
+          return {
+            ...validated,
+            persistentId: generatePersistentId({
+              creationTime: tLoggedCall.requestedAt,
+              key: `${tLoggedCall.id}`,
+              nodeId: node.id,
+            }),
+          };
+        } catch (e) {
+          console.error(e);
+          return null;
+        }
+      })
+      .filter(truthyFilter);
+
+    const inputDataChannel = await prisma.dataChannel.findFirst({
+      where: { destinationId: node.id },
+    });
+
+    if (!inputDataChannel) return;
+
+    const { datasetEntryInputsToCreate, datasetEntryOutputsToCreate, nodeEntriesToCreate } =
+      await prepareDatasetEntriesForImport({
+        projectId: node.projectId,
+        nodeId: node.id,
+        dataChannelId: inputDataChannel.id,
+        entriesToImport,
+      });
+
+    await prisma.$transaction([
+      prisma.datasetEntryInput.createMany({
+        data: datasetEntryInputsToCreate,
+        skipDuplicates: true,
+      }),
+      prisma.datasetEntryOutput.createMany({
+        data: datasetEntryOutputsToCreate,
+        skipDuplicates: true,
+      }),
+      prisma.nodeEntry.createMany({
+        data: nodeEntriesToCreate,
+        skipDuplicates: true,
+      }),
+    ]);
+  },
+};
+
+const calculateSampleRateHash = (sampleRate: number) => {
```

Comment:
So this is how I was expecting to implement this but I am still super low-confidence it will work lol. But we can test later I guess, none of this goes live immediately?

Reply:
Correct